### PR TITLE
Integrate XCM V1

### PR DESCRIPTION
**Pull request by tolak**
This PR contains all stuff of XCM V1 integration. Things we need to do are as follows:

- [x] Basic functionality implementation(xtransfer-pallets)
  - [x] mod matcher, a custom asset matcher
  - [x] mod assets(pallet), sibling parachain assets management
    - [x] assets registry
    - [x] non-reserve assets balance management
    - [x] non-reserve assets deposit/withdraw
  - [x] mod xcmtransfer(pallet), sibling parachain transfer based on xcm
    - [x] transfer reserve asset(PHA) to other sibling parachain
    - [x] receive non-reserve assets from other sibling parachain
    - [x] transfer non-reserve assets back to other sibling parachain
  - [x] multi-assets weight trader(*static multi-assets weight  trader is easy, havn't find a good way to impl a dynamic multi-assets weight trader so far*)
- [x] Unit test of xtransfer-pallets
- [x] Benchmark of xtransfer-pallets

### CODE REVIEW

File: pallets/xcm-transfer/src/xcm_transfer.rs [hunk 1]
```diff
@@ -0,0 +1,556 @@
...
+
+		ParaA::execute_with(|| {
+			// ParaA register it's own native asset
+			assert_ok!(XTransferAssets::register_asset(
+				ParaOrigin::root(),
+				b"ParaA Native Asset".to_vec(),
+				MultiLocation {
+					parents: 0,
+					interior: Here,
+				},
```
[the remark this answers is not in the document]
Comment (tolak) [reply pallets/xcm-transfer/src/xcm_transfer.rs]: Yes, more clear, already did like this

File: pallets/xcm-transfer/src/xcm_helper.rs [hunk 2]
```diff
@@ -78,10 +78,10 @@ pub mod xcm_helper {
 		pub fn is_native_asset_id(id: &MultiLocation) -> bool {
 			let native_locations = [
 				MultiLocation::here(),
-				MultiLocation {
-					parents: 1,
-					interior: X1(Parachain(T::get().into())),
-				},
+				(
+                    1,
```
Comment (h4x3rotab) [inline pallets/xcm-transfer/src/xcm_helper.rs:82]: indent. may need `cargo fmt`

File: pallets/xcm-transfer/src/assets.rs [hunk 3]
```diff
@@ -0,0 +1,367 @@
...
+				.ok_or(Error::<T>::AssetNotFound)?
+				.saturated_into();
+			let who = T::AccountIdConverter::convert_ref(who)
+				.map_err(|()| Error::<T>::AccountIdConversionFailed)?;
+			let balance_amount = amount
+				.try_into()
+				.map_err(|_| Error::<T>::AmountToBalanceConversionFailed)?;
+
+			if NativeAssetFilter::<T::ParachainInfo>::is_native_asset(&what) {
+				let _imbalance = T::Currency::deposit_creating(&who, balance_amount);
```
[the remark this answers is not in the document]
Comment (h4x3rotab) [reply pallets/xcm-transfer/src/assets.rs:240]: Turned out the creation and destroy is handled by xcm executor

**Review by h4x3rotab — APPROVED** [review]

**Comment by h4x3rotab** [comment]
Let's merge! Then we can look forward to migrate to xcm v2 (v3 is on the way though 😅)

**Pull request merged**